### PR TITLE
KAFKA-16437 - Upgrade to Jakarta and Jetty 12 (KIP-1032)

### PR DESCRIPTION
**Pull request by cshannon**
This commit implements the changes for [KIP-1032](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1032%3A+Upgrade+to+Jakarta+and+JavaEE+10+in+Kafka+4.0). This updates Kafka to Jakarta specs, JavaEE 10 and Jetty 12. The changes here primarily effect Kafka Connect and MM2.

**Todo/Notes:**

1. I bumped the connect modules to JDK 17 but I also had to bump a couple other things that had a dependency on conect. The tools project depends on connect so that had to be bumped, and streams depends on tools so that needed to be bumped. This means we may need to separate some things if we don't want to enforce JDK 17 on streams.
2. There is an issue with a test in DedicatedMirrorIntegrationTest that I had to change for now that involves escaping characters and not quite sure what to do about it yet. The cause is the Servlet 6 spec changing what is allowed in the path. See: https://github.com/jetty/jetty.project/issues/11890
3. I had to configure the idle timeout in Jetty requests to match our request timeout so tests didn't fail. This was needed to fix the `ConnectWorkerIntegrationTest#testPollTimeoutExpiry()` test

Testing is being done by just using the existing tests for Connect and MM2 which should be sufficient.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cshannon** [comment]
@gharris1727 and @C0urante - If either (or both) of you could take a look at some point and see what you think. There's no rush but I wanted to get this PR going now that trunk has been bumped to 4.0.0-SNAPSHOT. I am going to be out of town the rest of the week but I'll be back online middle of next week.

**Comment by cshannon** [comment]
The checks on Java 8 and Java 11 fail because this PR requires JDK 17+

File: build.gradle
Comment (gharris1727) [inline build.gradle]: Streams should still be built and tested with Java 11. I think this means we're blocked by https://issues.apache.org/jira/browse/KAFKA-17857

Also we are going to probably conflict with https://issues.apache.org/jira/browse/KAFKA-17811, and we can add connect and mirror to that list of modules to build with 17 as the minimum.

Comment (cshannon) [reply build.gradle]: I can just keep this PR open until we are unblocked and periodically update it by merging trunk into the PR to make sure I resolve any conflicts so not a huge deal.

**Comment by chia7712** [comment]
@cshannon Could you please rebase code to include #17522?

**Comment by cshannon** [comment]
@chia7712 - I merged the latest trunk and fixed the conflicts so this PR is now up to date

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (mumrah) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: nit: did you mean to add this space?

File: gradle/spotbugs-exclude.xml
Comment (mumrah) [inline gradle/spotbugs-exclude.xml]: This comment confused me at first. Can you include the description of `DP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED` to make it clear to the reader?

> Classloaders should only be created inside doPrivileged block

File: gradle/spotbugs-exclude.xml
Comment (mumrah) [inline gradle/spotbugs-exclude.xml]: Did any relevant code change in this PR for this spotbugs exclusion?

**Review by mumrah — COMMENTED** [review]
Thanks for the patch @cshannon. Overall, I think this change looks good. A few small things inline.

For big PRs like this, we rely heavily on our tests to ensure we didn't break anything. Can you run the test suite on this PR a few times just to see if anything shakes out? You can trigger a new build by pushing an empty commit (`git commit --allow-empty`)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (cshannon) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Nope, that was definitely inadvertent so I will fix that.

File: gradle/spotbugs-exclude.xml
Comment (cshannon) [reply gradle/spotbugs-exclude.xml]: I don't remember why I needed to add this because I first worked on this back in July, so I went back to take a look at the file and it looks like I can remove this block entirely because @chia7712 recently added the [same](https://github.com/apache/kafka/blob/0f2a96879030e1ccc575ed10b5cb113efa8200bb/gradle/spotbugs-exclude.xml#L550-L553) exclusion in https://github.com/apache/kafka/pull/17426

**Comment by cshannon** [comment]
@mumrah - Thanks for the review, I pushed an update and I believe I [addressed](https://github.com/apache/kafka/pull/16754/commits/b3f5c11675a5b9a3ffdfd0de47870746f3ac061b) all 3 of your comments (as well as merging in the latest from trunk) and that kicked off a new build. I can also run tests a couple more times with an empty commit after the tests finish.

**Review by mumrah — APPROVED** [review]
I re-ran the tests three times, looks good. I'm approving this now, but will wait a few days before merging to see if @gharris1727 or @C0urante (or @chia7712) have anything to add.

**Comment by cshannon** [comment]
> I re-ran the tests three times, looks good. I'm approving this now, but will wait a few days before merging to see if @gharris1727 or @C0urante (or @chia7712) have anything to add.

@mumrah - Sounds good, thanks for the review. 

I forgot to mention that I looked more into what to do about having to [change](https://github.com/apache/kafka/pull/16754/files#diff-ab24c09f7ecde101419a599c0ee3625a8f180911149a252792490a5618c7dc7cR283) the `DedicatedMirrorIntegrationTest` to prevent a failure because of the Servlet 6 and Jakarta EE10 spec change that no longer allows ambiguous characters anymore in the URL path. 

If we wanted to restore the previous behavior to allow it, we could [set](https://javadoc.jetty.org/jetty-12/org/eclipse/jetty/server/HttpConfiguration.html#setUriCompliance(org.eclipse.jetty.http.UriCompliance)) the UriComplianceMode to be [LEGACY](https://javadoc.jetty.org/jetty-12/org/eclipse/jetty/http/UriCompliance.html#LEGACY). See this comment: https://github.com/jetty/jetty.project/issues/11890#issuecomment-2156449534

However, I would say it's better to not allow it as the spec now says it's a violation so I think the current change I made is fine, but I guess it depends if there's use cases to allow it.

Edit: Here is the relevant part of the spec https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0.html#uri-path-canonicalization

File: gradle/dependencies.gradle
Comment (chia7712) [inline gradle/dependencies.gradle:168]: Could you please remove unnecessary item from license file after this change?

https://github.com/apache/kafka/blob/633cab91c1d5ca80763103fa38fb778dcf545709/LICENSE-binary#L305

Also, the following rule seems to be unnecessary.

https://github.com/apache/kafka/blob/trunk/build.gradle#L2900

File: gradle/dependencies.gradle
Comment (chia7712) [inline gradle/dependencies.gradle]: `jetty` 12.0.15 is using `jakarta.servlet` 6.0.0, so maybe we can remove this unused declaration?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java
Comment (chia7712) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java]: typo: `reqest` -> `request`

Comment (chia7712) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java]: This comment is valid to me, as the default idle timeout for HttpClient is 30 seconds, and the default request timeout is 90 seconds.

**Review by chia7712 — COMMENTED** [review]
@cshannon thanks for this patch. a couple of comments are left. PTAL

File: gradle/dependencies.gradle
Comment (cshannon) [reply gradle/dependencies.gradle]: Yeah it is being pulled in transitively by the ee10 Jetty jars so we can probably remove it because it's tied to that version.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java]: Could you keep most of these suspicious characters, and just remove the ones which are ambiguous?
From reading the linked issue, it might just be the `/` character, and the remainder are still valid.

It looks like this test was sensitive enough to find the underlying change in the library, so I'd like to keep it sensitive in case something like this happens again.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/util/SSLUtilsTest.java
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/util/SSLUtilsTest.java]: nit typo: bofore

**Review by gharris1727 — APPROVED** [review]
Thanks @cshannon for this change, thanks so much for getting this completed and working through the merge conflicts with all of the rest of the activity on trunk!

My comments are non-blocking, i'm happy to merge this as-is, but i'll let you address any other comments and merge tomorrow before feature freeze.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java
Comment (cshannon) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java]: So there's actually a few characters that are not allowed, control characters, backslash, etc. I went ahead and removed a few problem characters so it works now. I also updated the comments for the section. You can see the exact rules in the `Rejecting Suspicious Sequences` section as part of https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#uri-path-canonicalization

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java
Comment (cshannon) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java]: If we want to configure the idle timeout I guess the question is the best way to do so. I think we'd either need to pass it as an argument to the request() methods or add it to the AbstractConfig for RestClient so it has access to it when constructing the HttpClient. I guess there's also the question if it should always match the request timeout or it could be configured as a separate config. Maybe this could be done as a follow on task after this is merged.

**Comment by cshannon** [comment]
@gharris1727 and @chia7712 - I think I addressed all the comments so this can be merged. The only outstanding [one](https://github.com/apache/kafka/pull/16754#discussion_r1879098262) was if we wanted to set the idle timeout on the HttpClient to match the same thing as the request timeout (or even make it configurable separately) but I am still not quite sure if we should and I think that might be better as a follow on discussion and don't think it's a blocker. I can work on a PR for that depending on what we wanted to do (if anything)

**Comment by cshannon** [comment]
It seems the JDK 23 build is timing out but I'm not sure why, the issue could be unrelated to the tests themselves. The build is currently running again (and appears hung but I can't see why until it finishes it says). I pulled the logs from yesterday to figure out why it was hung and it shows:

```
2024-12-11T02:27:43.9906670Z Gradle Test Run :streams:integration-tests:test > Gradle Test Executor 183 > VersionedKeyValueStoreIntegrationTest > shouldManualUpgradeFromNonVersionedTimestampedToVersioned() STARTED
2024-12-11T02:28:31.2902594Z 
2024-12-11T02:28:31.2904162Z Gradle Test Run :streams:integration-tests:test > Gradle Test Executor 183 > VersionedKeyValueStoreIntegrationTest > shouldManualUpgradeFromNonVersionedTimestampedToVersioned() PASSED
2024-12-11T02:28:33.2903866Z Copy JUnit XML for integration-tests to /home/runner/work/kafka/kafka/build/junit-xml/streams/integration-tests/test
2024-12-11T03:05:56.8106484Z Timed out after 175 minutes. Dumping threads now...
```

That output shows it finishes tests and just does nothing for 37 minutes when trying to copy the XML files and then times out

**Comment by chia7712** [comment]
@cshannon There is a flaky test causing a timeout, and we are going to disable it temporarily. See [#18138](https://github.com/apache/kafka/pull/18138).

**Review by chia7712 — APPROVED** [review]

**Comment by chia7712** [comment]
This pull request causes a build issue. I have opened [KAFKA-18235](https://issues.apache.org/jira/browse/KAFKA-18235) to address and fix it.

**Pull request merged**
